File: lib/go/telepact/internal/binary/DecodeKeys.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package binary

import (
	"fmt"
	"strconv"
)

// DecodeKeys recursively decodes map keys using the provided BinaryEncoding lookup tables.
func DecodeKeys(given any, encoding *BinaryEncoding) (any, error) {
	switch typed := given.(type) {
	case map[string]any:
		result := make(map[string]any, len(typed))
		for key, value := range typed {
			decodedValue, err := DecodeKeys(value, encoding)
			if err != nil {
				return nil, err
			}
			if decodedKey, ok, convErr := decodeNumericStringKey(key, encoding); convErr != nil {
				return nil, convErr
			} else if ok {
				result[decodedKey] = decodedValue
				continue
			}
			result[key] = decodedValue
		}
		return result, nil
	case map[int]any:
		result := make(map[string]any, len(typed))
		for key, value := range typed {
			decodedValue, err := DecodeKeys(value, encoding)
			if err != nil {
				return nil, err
			}
			newKey, ok := encoding.DecodeMap[key]
			if !ok {
				return nil, NewBinaryEncodingMissing(key)
			}
			result[newKey] = decodedValue
		}
		return result, nil
	case map[any]any:
		result := make(map[string]any, len(typed))
		for rawKey, value := range typed {
			decodedValue, err := DecodeKeys(value, encoding)
			if err != nil {
				return nil, err
			}

			switch key := rawKey.(type) {
			case string:
				if decodedKey, ok, convErr := decodeNumericStringKey(key, encoding); convErr != nil {
					return nil, convErr
				} else if ok {
					result[decodedKey] = decodedValue
					break
				}
				result[key] = decodedValue
			default:
				intKey, ok := toInt(key)
				if !ok {
					return nil, fmt.Errorf("invalid key type for decode: %T", rawKey)
				}
				newKey, exists := encoding.DecodeMap[intKey]
				if !exists {
					return nil, NewBinaryEncodingMissing(intKey)
				}
				result[newKey] = decodedValue
			}
		}
		return result, nil
	case []any:
		result := make([]any, len(typed))
		for i, item := range typed {
			decodedValue, err := DecodeKeys(item, encoding)
			if err != nil {
				return nil, err
			}
			result[i] = decodedValue
		}
		return result, nil
	default:
		return given, nil
	}
}

func decodeNumericStringKey(key string, encoding *BinaryEncoding) (string, bool, error) {
	if encoding == nil {
		return "", false, nil
	}
	if len(key) == 0 {
		return "", false, nil
	}
	if key[0] == '-' || (key[0] >= '0' && key[0] <= '9') {
		if intKey, err := strconv.Atoi(key); err == nil {
			if decoded, ok := encoding.DecodeMap[intKey]; ok {
				return decoded, true, nil
			}
			return "", false, NewBinaryEncodingMissing(intKey)
		}
	}
	return "", false, nil
}

File: lib/py/telepact/internal/binary/DecodeKeys.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING
from typing import Callable
from ...internal.binary.BinaryEncodingMissing import BinaryEncodingMissing

if TYPE_CHECKING:
    from ...internal.binary.BinaryEncoding import BinaryEncoding


def _decode_keys_recursive(value: object, decode_key: 'Callable[[object], str]') -> object:
    value_type = type(value)

    if value_type is dict:
        decoded: dict[str, object] = {}
        for key, item in value.items():
            if type(key) is str:
                decoded_key = key
            else:
                decoded_key = decode_key(key)
            decoded[decoded_key] = _decode_keys_recursive(item, decode_key)
        return decoded
    if value_type is list:
        return [_decode_keys_recursive(item, decode_key) for item in value]
    return value


def decode_keys(given: object, binary_encoder: 'BinaryEncoding') -> object:
    decode_table = binary_encoder.decode_table

    def decode_key(key: object) -> str:
        try:
            return decode_table[key]
        except (IndexError, TypeError) as exc:
            raise BinaryEncodingMissing(key) from exc

    if isinstance(given, dict):
        return _decode_keys_recursive(given, decode_key)
    if isinstance(given, list):
        return _decode_keys_recursive(given, decode_key)
    return given
